### ISSUE TITLE
Encountered weird drop in performance when running a function

### ISSUE THREAD

**Issue by polakowo** [numba/numba#5743]
Given the following code:

```py
from numba import njit, f8
import numpy as np

@njit
def main_func_nb(a, func_nb, *args):
    b = np.empty_like(a, dtype=f8)

    for col in range(a.shape[1]):
        for i in range(a.shape[0]):
            b[i, col] = func_nb(col, i, *args)
    return b


@njit
def func_nb(col, i, x1, x2, x3, is_something):
    if is_something:
        return x1[i, col] + x2[i, col] + x3[i, col] + 1
    else:
        return x1[i, col] + x2[i, col] + x3[i, col]
    
a = x1 = x2 = x3 = np.random.uniform(size=(1000, 1000))

%timeit _ = main_func_nb(a, func_nb, x1, x2, x3, False)
7.34 ms ± 199 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

If I move the `1` in `func_nb` before any of the `x` arrays, the performance drops twentyfold.

```py
@njit
def func_nb(col, i, x1, x2, x3, is_something):
    if is_something:
        return 1 + x1[i, col] + x2[i, col] + x3[i, col]
    else:
        return x1[i, col] + x2[i, col] + x3[i, col]

%timeit _ = main_func_nb(price, func_nb, x1, x2, x3, False)
116 ms ± 3.82 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I can't reproduce this behavior if there is less than three input arrays, or if `is_something` check is dropped, or if contents of `func_nb` is moved to `main_func_nb`.

Any thoughts?

**Comment by esc** [comment]
@polakowo thanks for submitting this to the Numba issue tracker. I can indeed reproduce the slow down that you describe:

```
In [4]: %timeit _ = main_func_nb(a, func_nb_fast, x1, x2, x3, False)
9.01 ms ± 66.2 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit _ = main_func_nb(a, func_nb_slow, x1, x2, x3, False)
67.6 ms ± 2.54 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Given that you already have a workaround, I would suggest to use `inspect_disasm_cfg()` on the `func_nb` variants and compare the output to figure out what happened. Without actually looking into the details, I would assume that the array expression `return 1 + x1[i, col] + x2[i, col] + x3[i, col]
` isn't being optimized and instead multiple temporaries are being created in memory, but that is just a guess.

**Comment by stuartarchibald** [comment]
Following a quick look at the machine code, think it's reference counting getting in the way.

**Comment by polakowo** [comment]
I’m not very experienced with Numba, so reading machine code wouldn’t help me resolve this issue. I submitted this because my vectorbt project depends a lot upon passing functions as arguments and unpacking tuples with large arrays, but this yields an unstable performance from time to time. The example above is just a simplified case.

**Comment by esc** [comment]
@polakowo thanks for following up on this. Question: is this a recent change that you observed? I.e. has this worked better with previous versions of Numba?

**Comment by esc** [comment]
I debugged this some more and this issue here may have the same root cause as: https://github.com/numba/numba/issues/5746

Specifically, the commit https://github.com/numba/numba/commit/8c74571a2f97cede08c491b6c018ed144072e8bd causes the issue. And it was a part of https://github.com/numba/numba/pull/5338

If I revert the 8c7457 on master, I no longer observe the performance hit:

```
💥 zsh» git --no-pager diff
diff --git i/numba/core/lowering.py w/numba/core/lowering.py
index 04e1ed68e9..a8da51e710 100644
--- i/numba/core/lowering.py
+++ w/numba/core/lowering.py
@@ -7,7 +7,7 @@ from llvmlite.llvmpy.core import Constant, Type, Builder

 from numba import _dynfunc
 from numba.core import (typing, utils, types, ir, debuginfo, funcdesc,
-                        generators, config, ir_utils, cgutils)
+                        generators, config, ir_utils, cgutils, removerefctpass)
 from numba.core.errors import (LoweringError, new_error_context, TypingError,
                                LiteralTypingError, UnsupportedError)
 from numba.core.funcdesc import default_mangler
@@ -204,6 +204,12 @@ class BaseLower(object):
                 print(self.module)
             print('=' * 80)

+        # Special optimization to remove NRT on functions that do not need it.
+        if self.context.enable_nrt and self.generator_info is None:
+            removerefctpass.remove_unnecessary_nrt_usage(self.function,
+                                                         context=self.context,
+                                                         fndesc=self.fndesc)
+
         # Run target specific post lowering transformation
         self.context.post_lowering(self.module, self.library)

💥 zsh» cat foo066.py
from numba import njit, f8
import numpy as np


@njit
def main_func_nb(a, func_nb, *args):
    b = np.empty_like(a, dtype=f8)

    for col in range(a.shape[1]):
        for i in range(a.shape[0]):
            b[i, col] = func_nb(col, i, *args)
    return b


@njit
def func_nb_fast(col, i, x1, x2, x3, is_something):
    if is_something:
        return x1[i, col] + x2[i, col] + x3[i, col] + 1
    else:
        return x1[i, col] + x2[i, col] + x3[i, col]


@njit
def func_nb_slow(col, i, x1, x2, x3, is_something):
    if is_something:
        return 1 + x1[i, col] + x2[i, col] + x3[i, col]
    else:
        return x1[i, col] + x2[i, col] + x3[i, col]


a = x1 = x2 = x3 = np.random.uniform(size=(1000, 1000))
💥 zsh» ipython
Python 3.8.2 (default, Mar 26 2020, 10:43:30)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: run foo066.py

In [2]: %timeit _ = main_func_nb(a, func_nb_fast, x1, x2, x3, False)
7.2 ms ± 557 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: %timeit _ = main_func_nb(a, func_nb_slow, x1, x2, x3, False)
6.62 ms ± 487 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by esc** [comment]
@polakowo probably the best temporary workaround we can offer for the time being is to use Numba 0.48.0 until this issue is fixed.

**Comment by polakowo** [comment]
I have done some tests on a more complex example with numba 0.48.0 and 0.49.1, here are the results.

Function as argument and tuple unpacking yields worst performance in both 0.48 and newest version, although the 0.48 does here better:
```py
from numba import njit, f8
import numpy as np

@njit
def simulate_nb(target_shape, init_capital, order_func_nb, *args):
    trade_size = np.full(target_shape, np.nan, dtype=f8)
    trade_price = np.full(target_shape, np.nan, dtype=f8)
    trade_fees = np.full(target_shape, np.nan, dtype=f8)
    cash = np.empty(target_shape, dtype=f8)
    shares = np.empty(target_shape, dtype=f8)

    for col in range(target_shape[1]):
        run_cash = init_capital
        run_shares = 0.

        for i in range(target_shape[0]):
            order_size, order_price, order_fees, order_fixed_fees, order_slippage = \
                order_func_nb(col, i, run_cash, run_shares, *args)
            if order_size != 0.:
                trade_size[i, col] = order_size
                trade_price[i, col] = order_price
                trade_fees[i, col] = order_fees * order_size * order_price + order_fixed_fees
            run_cash += -order_size * order_price
            run_shares += order_size
            cash[i, col], shares[i, col] = run_cash, run_shares

    return trade_size, trade_price, trade_fees, cash, shares


@njit(cache=True)
def signals_order_func_nb(col, i, run_cash, run_shares, entries, exits, size, entry_price,
                          exit_price, fees, fixed_fees, slippage, accumulate):
    order_size = 0.
    order_price = 0.
    if entries[i, col] and not exits[i, col]:
        if run_shares == 0. or accumulate:
            order_size = abs(size[i, col])
            order_price = entry_price[i, col]
    elif not entries[i, col] and exits[i, col]:
        # Sell everything
        if run_shares > 0.:
            order_size = -np.inf
            order_price = exit_price[i, col]
    return order_size, order_price, fees[i, col], fixed_fees[i, col], slippage[i, col]

target_shape = (1000, 1000)
init_capital = 1000
entries = np.random.randint(0, 2, size=target_shape).astype(np.bool)
exits = np.random.randint(0, 2, size=target_shape).astype(np.bool)
size = np.random.uniform(-10, 10, size=target_shape)
entry_price = exit_price = np.random.uniform(1, 10, size=target_shape)
fees = np.full(target_shape, 0.01)
fixed_fees = np.full(target_shape, 1)
slippage = np.full(target_shape, 0.01)
accumulate = False

%timeit simulate_nb(\
    target_shape, init_capital, signals_order_func_nb, \
    entries, exits, size, entry_price,\
    exit_price, fees, fixed_fees, slippage, accumulate)
```

```
# 0.48.0
155 ms ± 4.75 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# 0.49.1
321 ms ± 49.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Passing arrays as positional arguments cuts processing time significantly in 0.48, in the newest version it's much slower (maybe due to the reason stated in previous comments):
```py
@njit
def simulate2_nb(target_shape, init_capital, order_func_nb, entries, exits, size, entry_price,
                          exit_price, fees, fixed_fees, slippage, accumulate):
    trade_size = np.full(target_shape, np.nan, dtype=f8)
    trade_price = np.full(target_shape, np.nan, dtype=f8)
    trade_fees = np.full(target_shape, np.nan, dtype=f8)
    cash = np.empty(target_shape, dtype=f8)
    shares = np.empty(target_shape, dtype=f8)

    for col in range(target_shape[1]):
        run_cash = init_capital
        run_shares = 0.

        for i in range(target_shape[0]):
            order_size, order_price, order_fees, order_fixed_fees, order_slippage = \
                signals_order_func2_nb(col, i, run_cash, run_shares, entries, exits, size, entry_price,
                          exit_price, fees, fixed_fees, slippage, accumulate)
            if order_size != 0.:
                trade_size[i, col] = order_size
                trade_price[i, col] = order_price
                trade_fees[i, col] = order_fees * order_size * order_price + order_fixed_fees
            run_cash += -order_size * order_price
            run_shares += order_size
            cash[i, col], shares[i, col] = run_cash, run_shares

    return trade_size, trade_price, trade_fees, cash, shares


@njit(cache=True)
def signals_order_func2_nb(col, i, run_cash, run_shares, entries, exits, size, entry_price,
                          exit_price, fees, fixed_fees, slippage, accumulate):
    order_size = 0.
    order_price = 0.
    if entries[i, col] and not exits[i, col]:
        if run_shares == 0. or accumulate:
            order_size = abs(size[i, col])
            order_price = entry_price[i, col]
    elif not entries[i, col] and exits[i, col]:
        # Sell everything
        if run_shares > 0.:
            order_size = -np.inf
            order_price = exit_price[i, col]
    return order_size, order_price, fees[i, col], fixed_fees[i, col], slippage[i, col]

%timeit simulate2_nb(\
    target_shape, init_capital, signals_order_func2_nb, \
    entries, exits, size, entry_price,\
    exit_price, fees, fixed_fees, slippage, accumulate)
```

```
# 0.48.0
21.9 ms ± 511 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
# 0.49.1
198 ms ± 10.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Now the fastest in 0.49.0 with indexing done in the main function:
```py
@njit
def simulate3_nb(target_shape, init_capital, entries, exits, size, entry_price,
                          exit_price, fees, fixed_fees, slippage, accumulate):
    trade_size = np.full(target_shape, np.nan, dtype=f8)
    trade_price = np.full(target_shape, np.nan, dtype=f8)
    trade_fees = np.full(target_shape, np.nan, dtype=f8)
    cash = np.empty(target_shape, dtype=f8)
    shares = np.empty(target_shape, dtype=f8)

    for col in range(target_shape[1]):
        run_cash = init_capital
        run_shares = 0.

        for i in range(target_shape[0]):
            order_size, order_price, order_fees, order_fixed_fees, order_slippage = \
                signals_order_func3_nb(col, i, run_cash, run_shares, entries[i, col], exits[i, col], 
                        size[i, col], entry_price[i, col], exit_price[i, col], fees[i, col], fixed_fees[i, col], 
                        slippage[i, col], accumulate)
            if order_size != 0.:
                trade_size[i, col] = order_size
                trade_price[i, col] = order_price
                trade_fees[i, col] = order_fees * order_size * order_price + order_fixed_fees
            run_cash += -order_size * order_price
            run_shares += order_size
            cash[i, col], shares[i, col] = run_cash, run_shares

    return trade_size, trade_price, trade_fees, cash, shares


@njit(cache=True)
def signals_order_func3_nb(col, i, run_cash, run_shares, entries, exits, size, entry_price,
                          exit_price, fees, fixed_fees, slippage, accumulate):
    order_size = 0.
    order_price = 0.
    if entries and not exits:
        if run_shares == 0. or accumulate:
            order_size = abs(size)
            order_price = entry_price
    elif not entries and exits:
        # Sell everything
        if run_shares > 0.:
            order_size = -np.inf
            order_price = exit_price
    return order_size, order_price, fees, fixed_fees, slippage

%timeit simulate3_nb(\
    target_shape, init_capital, \
    entries, exits, size, entry_price,\
    exit_price, fees, fixed_fees, slippage, accumulate)
```

```
# 0.48.0
33.9 ms ± 542 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
# 0.49.1
34.4 ms ± 977 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I'm mainly interested in the first version as it allows defining some flexible code such as rolling windows with custom apply function, but you see that even 0.48 has issues here.

**Comment by stuartarchibald** [comment]
Root cause is the same as this https://github.com/numba/numba/issues/5746

**Comment by stuartarchibald** [comment]
Fixed in #5795
